File: ga5/imdb_scraper.js
// Use puppeteer-extra to add plugins
const puppeteer = require('puppeteer-extra');
// Add the stealth plugin
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());

async function scrapeImdbRatings(minRating = 7.0, maxRating = 8.0, count = 25) {
    let browser;
    try {
        console.log('Launching stealth browser...');
        browser = await puppeteer.launch({
            headless: true,
            args: ['--no-sandbox', '--disable-setuid-sandbox']
        });
        const page = await browser.newPage();
        await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36');

        const url = `https://www.imdb.com/search/title/?user_rating=${minRating},${maxRating}`;
        console.log(`Navigating to ${url}...`);
        await page.goto(url, { waitUntil: 'networkidle2' });

        // A more generic selector that is more likely to be present
        const listSelector = "li.ipc-metadata-list-summary-item";
        console.log(`Waiting for selector: ${listSelector}`);
        await page.waitForSelector(listSelector, { timeout: 20000 });

        // Extract the data
        const movies = await page.evaluate((count) => {
            const results = [];
            const items = document.querySelectorAll("li.ipc-metadata-list-summary-item");
            
            for (let i = 0; i < items.length && i < count; i++) {
                const item = items[i];
                try {
                    const titleLink = item.querySelector('a.ipc-title-link-wrapper');
                    const titleText = titleLink.querySelector('h3.ipc-title__text').innerText || '';
                    const title = titleText.replace(/^\d+\.\s*/, '').trim();

                    const href = titleLink.getAttribute('href');
                    const idMatch = href.match(/\/title\/(tt\d+)\//);
                    const id = idMatch ? idMatch[1] : null;

                    const metadataItems = item.querySelectorAll('.sc-b189961a-8');
                    const year = metadataItems[0] ? metadataItems[0].innerText : null;
                    
                    const ratingElement = item.querySelector('span.ipc-rating-star');
                    const rating = ratingElement ? ratingElement.innerText.split(/\s/)[0] : null;

                    if (id && title && year && rating) {
                        results.push({ id, title, year, rating });
                    }
                } catch (e) {
                    // Ignore malformed items
                }
            }
            return results;
        }, count);

        return JSON.stringify(movies, null, 2);

    } catch (error) {
        console.error('An error occurred during scraping:', error);
        return null;
    } finally {
        if (browser) {
            await browser.close();
            console.log('Browser closed.');
        }
    }
}

// --- Main execution ---
(async () => {
    const movieDataJson = await scrapeImdbRatings();
    if (movieDataJson) {
        console.log(movieDataJson);
    }
})();

File: ga5/imdb_scraper.py
import json
import re
import time
from selenium import webdriver
from selenium.webdriver.firefox.service import Service as FirefoxService
from selenium.webdriver.firefox.options import Options as FirefoxOptions
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import geckodriver_autoinstaller
from bs4 import BeautifulSoup

def scrape_imdb_with_firefox(min_rating=7.0, max_rating=8.0, count=25):
    """
    Scrapes IMDb using Selenium with Firefox to handle dynamic content.

    Args:
        min_rating (float): The minimum IMDb user rating.
        max_rating (float): The maximum IMDb user rating.
        count (int): The maximum number of titles to return.

    Returns:
        str: A JSON formatted string containing the movie data, or None.
    """
    # Configure Firefox options for headless operation
    firefox_options = FirefoxOptions()
    firefox_options.add_argument("--headless")
    firefox_options.add_argument("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0")

    driver = None
    try:
        # Automatically install and set up geckodriver
        geckodriver_autoinstaller.install()
        
        driver = webdriver.Firefox(options=firefox_options)
        
        url = f"https://www.imdb.com/search/title/?user_rating={min_rating},{max_rating}"
        driver.get(url)

        # Wait for the page to load and try to accept cookies
        wait = WebDriverWait(driver, 20)
        try:
            # A more generic way to find consent buttons
            accept_button = wait.until(EC.element_to_be_clickable((By.XPATH, "//button[contains(text(), 'Accept')]")))
            accept_button.click()
        except Exception:
            pass # Continue if no button is found

        # Wait for a more generic list container to be present
        wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, ".ipc-metadata-list")))
        
        html_content = driver.page_source
        
        soup = BeautifulSoup(html_content, 'html.parser')
        
        movie_list = soup.find_all('li', class_='ipc-metadata-list-summary-item')
        if not movie_list:
            print("Could not find any movie items in the list.")
            return None

        extracted_data = []
        id_regex = re.compile(r'/title/(tt\d+)/')

        for item in movie_list[:count]:
            try:
                title_link = item.find('a', class_='ipc-title-link-wrapper')
                href = title_link['href']
                match = id_regex.search(href)
                imdb_id = match.group(1) if match else None

                title_text = title_link.find('h3').text
                title = re.sub(r'^\d+\.\s*', '', title_text).strip()

                metadata_div = item.find('div', class_='sc-b189961a-7')
                metadata_spans = metadata_div.find_all('span', class_='sc-b189961a-8')
                year = metadata_spans[0].text.strip() if len(metadata_spans) > 0 else None
                
                rating_span = item.find('span', class_='ipc-rating-star')
                rating = rating_span.text.split()[0].strip() if rating_span else None

                if all([imdb_id, title, year, rating]):
                    extracted_data.append({
                        "id": imdb_id,
                        "title": title,
                        "year": year,
                        "rating": rating
                    })
            except (AttributeError, IndexError):
                continue
        
        return json.dumps(extracted_data, indent=2)

    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        return None
    finally:
        if driver:
            driver.quit()

if __name__ == "__main__":
    movie_data_json = scrape_imdb_with_firefox()
    if movie_data_json:
        print(movie_data_json)